### PR TITLE
Implement FixedCapacityVector.

### PR DESCRIPTION
**Pull request by markdroth**
This provides better C++-style semantics than C-style arrays, which are one of the main areas where we suffer from the problem described in #14219.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by vjpai** [comment]
Please rebase; I think there was a master sanity failure when you sent in this PR. Thanks!

**Comment by markdroth** [comment]
Done.  PTAL.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/gprpp/fixed_capacity_vector.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+// supported by STL containers (e.g., iterators).  Those can be added as
+// needed.
+
+template <typename T>
+class FixedCapacityVector {
+ public:
+  static UniquePtr<FixedCapacityVector> Create(size_t capacity) {
+    // Allocate space for both this object and the elements.
+    FixedCapacityVector* v = static_cast<FixedCapacityVector*>(
+        gpr_malloc(sizeof(FixedCapacityVector) + (capacity * sizeof(T))));
```
Comment (vjpai) [inline src/core/lib/gprpp/fixed_capacity_vector.h]: So FixedCapacityVector is likely to have size 8, which would make the actual data located here to be on an offset of 8 relative to the 16-aligned block returned by malloc. Is there any situation where the actual data array needs to be 16-aligned?

Comment (markdroth) [reply src/core/lib/gprpp/fixed_capacity_vector.h]: Good point.  Fixed.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by markdroth** [comment]
I realized that I can accomplish this same thing in a slightly cleaner way by implementing the `reserve()` method on `InlinedVector`, which I've done in #14845.  Reverting this PR in favor of that one.

**Pull request closed without merging**